instaloader/instaloadercontext.py

```diff
@@ -521,13 +521,18 @@ def get_raw(self, url: str, _attempt=1) -> requests.Response:
                 raise QueryReturnedNotFoundException("404 when accessing {}.".format(url))
             raise ConnectionException("HTTP error code {}.".format(resp.status_code))
 
-    def get_and_write_raw(self, url: str, filename: str) -> None:
+    def get_and_write_raw(self, url: str, filename: str, get_extension: bool = False) -> str:
         """Downloads and writes anonymously-requested raw data into a file.
 
         :raises QueryReturnedNotFoundException: When the server responds with a 404.
         :raises QueryReturnedForbiddenException: When the server responds with a 403.
         :raises ConnectionException: When download repeatedly failed."""
-        self.write_raw(self.get_raw(url), filename)
+        resp = self.get_raw(url)
+        if get_extension and 'Content-Type' in resp.headers and resp.headers['Content-Type']:
+            extension = '.' + resp.headers['Content-Type'].split(';')[0].split('/')[-1].lower().replace('jpeg', 'jpg')
+            filename += extension
```

Comment:
It would be probably much easier to have this in `download_pic()` and then let it call `get_raw()` and `write_raw()` rather than `get_and_write_raw()`, since we'd have to check whether a file with the appropriate extension already exists.